compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -390,45 +475,195 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
-    SmallVector<OpFoldResult> resultOffsets, resultSizes;
-    if (failed(op.getResultTilePosition(rewriter, result.getResultNumber(),
-                                        offsets, sizes, resultOffsets,
-                                        resultSizes))) {
-      return rewriter.notifyMatchFailure(
-          op, "failed to rewrite destructive update");
+  std::swap(candidateSliceOps, remainingCandidates);
+  return slicesFromSameProducer;
+}
+
+FailureOr<TileAndFuseResult>
+TileAndFuseDispatchUsingSCFForOp::returningMatchAndRewrite(
+    TilingInterface op, PatternRewriter &rewriter) const {
+  TileAndFuseResult tileAndFuseResult;
+  // Apply the tiling pattern.
+  FailureOr<TilingResult> tilingResult =
+      tilingPattern.returningMatchAndRewrite(op, rewriter);
+  if (failed(tilingResult)) {
+    return failure();
+  }
+  // If there is no tiling then there is nothing to do for fusion.
+  if (!tilingResult->tiledLoops.any()) {
+    return tileAndFuseResult;
+  }
+  tileAndFuseResult.tiledAndFusedOps.push_back(tilingResult->tiledOp);
+  tileAndFuseResult.loops.append(tilingResult->loops);
+
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(tilingResult->tiledOp);
+  // 1. Collect the operands of the operation that are slices. Use a worklist
+  //    to tile and fuse the entire DAG.
+  SmallVector<tensor::ExtractSliceOp> candidateSliceOps;
+  addCandidateSlices(tileAndFuseResult.tiledAndFusedOps.back(),
+                     candidateSliceOps);
+
+  // Iterate over worklist till all slices are fused.
+  while (!candidateSliceOps.empty()) {
+    // 2a. All slices from the same operations are processed concurrently. Find
+    //     the slices that are from the same operation. Prune these slices from
+    //     current candidate list.
+    SmallVector<tensor::ExtractSliceOp> currSliceOps =
+        getSlicesFromSameProducer(candidateSliceOps);
+    if (currSliceOps.empty()) continue;
+    auto producerVal = currSliceOps.front().getSource().cast<OpResult>();
+    TilingInterface producer =
+        currSliceOps.front().getSource().getDefiningOp<TilingInterface>();
+    if (!producer) {
+      continue;
     }
-    SmallVector<OpFoldResult> resultStrides(resultOffsets.size(),
-                                            rewriter.getIndexAttr(1));
 
-    SmallVector<OpFoldResult> combinedOffsets, combinedSizes, combinedStrides;
-    if (failed(IREE::Flow::foldOffsetsSizesAndStrides(
-            rewriter, loc, storeOp.getMixedOffsets(), storeOp.getMixedSizes(),
-            storeOp.getMixedStrides(), storeOp.getDroppedDims(), resultOffsets,
-            resultSizes, resultStrides, combinedOffsets, combinedSizes,
-            combinedStrides))) {
-      return rewriter.notifyMatchFailure(
-          op, "failed to create tiled flow.dispatch.tensor.store op");
+    tensor::ExtractSliceOp sliceOp = currSliceOps.back();
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(sliceOp);
```

Comment:
Since the tiled producer will be used by multiple consumers, is it guaranteed that inserting at the last slice op will dominate all the uses?